text/0000-feature-metadata.md

```diff
@@ -0,0 +1,296 @@
+
+-   `requires`: This is synonymous with the existing array describing required
+    features. For example, `foo = ["dep:serde", "otherfeat"]` will be identical to
+    `foo = { requires = ["dep:serde", "otherfeat"] }`
+-   `doc`: A markdown docstring describing the feature. Like with `#[doc(...)]`,
+    the first line will be treated as a summary.
+-   `deprecated`: This can be either a simple boolean, a string, or an object
+    with `since` and/or `note` keys. Cargo will warn downstream crates using this
+-   `public`: A boolean flag defaulting to `true` that indicates whether or not
+    downstream crates should be allowed to use this feature
+-   `unstable`: A boolean flag indicating that the feature is only usable with
+    nightly Rust, or is otherwise not API-stable
+
+If a downstream crate attempts to use the features `baz` and `qux`, they will
+see messages like the following:
+
+```
+warning: feature `quux` on crate `mycrate` is deprecated since version
+  1.2.3: "don't use this!"
+
+error: feature `baz` on crate `mycrate` is private and cannot be used by
+  downstream crates
+```
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Validation and parsing of the new schema, described above, shoud be relatively
+straightforward. Each of the added keys is discussed individually in the
+following sections:
+
+## `requires`
+
+`requires` will take the place of the feature dependency array that currently
+exists. Semantics will remain unchanged.
+
+This is a required key.
+
+## `doc`
+
+`doc` is the most straightforward: it accepts markdown-flavored text, and should
+be thought of as the equivalent to a `#[doc(...)]` attribute. Like doc comments,
+the first line should be usable as a summary, though most use cases will only
+have one line.
+
+There is nothing in this RFC that cargo `must` do with this action, since it is
+mainly intended for the consumption of `rustdoc` or `docs.rs`. However, it can
+be used for general diagnostic information.
+
+## `deprecated`
+
+`deprecated` should be thought of as the equivalent of the [`deprecated`]
+attribute in Rust source. The value can be a boolean, string, or an object with
+`since` or `note` keys. Schema rules are as follows:
+
+-   If a boolean value, `false` indicates not deprecated and `true` indicates
+    deprecated
+-   If an object, the keys `since` and/or `note` can be specified
+-   An empty object is not allowed to avoid the ambiguous
+    `foo = { deprecated = {} }`
+-   A string `foo = { deprecated = "my msg" }` will be equivalent to if that string
+    was specified in the `note` field:
+    `foo = { deprecated = { note = "my msg" } }`
+-   If not specified, the default is `false`
+
+If a downstream crate attempts to use a feature marked `deprecated`, Cargo
+should produce a warning. There are two exceptions to this:
+
+-   If a `since` is specified that is later than the crate's current version,
+    this warning should not be emitted.
```

Comment:
"current version" isn't accurate

For the deprecation warning to show up, your lockfile most likely has a `since` version in it, unless someone did `since` in the future

What I was suggesting was to check if `since` is newer than the minimum version that matches the version requirement

Reply:
@epage are you suggesting to prefer an older version over a newer one, if an enabled feature is deprecated in the newer version but not in the older one?

I do not like this, for two reasons:

- If a feature is deprecated, that probably means that it will be removed in the future; it does **not** mean that it is broken and should not be used. I don't see a good reason for preferring an outdated version just to silence a deprecation warning.

- Rust ignores `since` field in the `#[deprecated]` attribute, it is only for documentation and better error messages. I think cargo should do the same here; the `since` field should not affect behavior of dependency resolution.

---

@tgross35 IMO specifying a newer version in the `since` field than the current version should just be an error.

Reply:
No, I'm suggesting we only report the warning if the minimum version that matches the version requirement is at least `since`.  As I said, we should avoid reporting a warning that is not actionable.

Reply:
Deprecations should be actionable, i.e. it should be possible to remove the feature or replace it with another one. If it's not clear, the deprecation should have a `note = "..."` explaining how it can be fixed.

Reply:
From your comment, I'm assuming you posted before seeing my edit.  I'd recommend going back over my comment

Reply:
> Say regex's `unicode` feature was deprecated in 1.100, replaced with `unicode2`. My dependency on regex is `regex = { version = "1.95.0", features = ["unicode"] }`. If I were given the deprecation warning, that means my `Cargo.lock` is pointing to 1.100. I could then replace `unicode` with the `unicode2` and the deprecation warning is gone. I publish my crate. Dependents of my crate will then break because their `Cargo.lock` is pointing to `1.96.0`.

This is exactly how `#[deprecated]` works today, if this is going to work differently then that should likely also change.

It would be difficult to accurately check this when more complicated version constraints are involved, imagine the version constraint `>=1.1.2, <3.0.0` with a feature that has `since = "1.1.0"` in the v1 branch and `since = "2.2.0"` in the v2 branch and `2.3.0` is currently locked, the metadata of the used version of the crate doesn't tell you the minimum v1 requirement is correct, you would just see that `1.1.2 < 2.2.0` and give a warning.

Reply:
Does checking a crate that uses `regex = { version = "1.95.0", features = ["unicode2"] }` do any sort of verification that `unicode2` is available for all possible dependency versions if it is locked at `1.100`? I don't believe so but I also haven't tested.

The line that this review thread references was just meant so that a crate author could do something like mark a few features `deprectated { since = "1.0.0" }` even if current crate version is < 1, and have them automatically start warning whenever that version is actually released. But I will eliminate this, since consistency with `#[deprecated]` is more important

Reply:
> This is exactly how #[deprecated] works today, if this is going to work differently then that should likely also change.

The fact that I wanted to do this for `#[deprecated]` is what inspired my comment

> It would be difficult to accurately check this when more complicated version constraints are involved, imagine the version constraint >=1.1.2, <3.0.0 with a feature that has since = "1.1.0" in the v1 branch and since = "2.2.0" in the v2 branch and 2.3.0 is currently locked, the metadata of the used version of the crate doesn't tell you the minimum v1 requirement is correct, you would just see that 1.1.2 < 2.2.0 and give a warning.

We could just check what is the lowest version in the index for a given version range like we do with rust-lang/cargo#5657

While there are corner cases where the logic might not work (like parallel development on two different branches **when** using complicated version requirements), I feel like that is enough of a corner case that showing false positive in that case is better than showing false positives in all of the cases where the version req is too low.

I am also fine with moving this to a future possibility so as to not derail this proposal with this discussion

Reply:
I added a comment about how the exact semantics of `since` are deferred and optional for `deprecated`: https://github.com/rust-lang/rfcs/pull/3416/files#diff-b1b8fe9a7278ab777a7fa13af204b0717b1ab818716d76854ed9617569ce08d9R145-R155

Reply:
Moved to #3486